api/v1alpha3/cluster_types.go

```diff
@@ -183,3 +186,43 @@ type ClusterList struct {
 func init() {
 	SchemeBuilder.Register(&Cluster{}, &ClusterList{})
 }
+
+// FailureDomains is a slice of FailureDomain objects.
+type FailureDomains []FailureDomain
+
+// GetByID returns the FailureDomain with the identifier supplied, if none is found returns nil.
+func (v FailureDomains) GetByID(id string) *FailureDomain {
+	for _, fd := range v {
+		if fd.ID == id {
+			return &fd
+		}
+	}
+	return nil
+
+}
+
+// FilterControlPlane returns a FailureDomain slice containing only the domains suitable to be used
+// for control plane nodes.
+func (v FailureDomains) FilterControlPlane(id string) (res FailureDomains) {
+	for _, fd := range v {
+		if fd.ControlPlane {
+			res = append(res, fd)
+		}
+	}
+	return
+}
+
+// FailureDomain is the Schema for Cluster API failure domains.
+// It allows controllers to understand how many failure domains a cluster can optionally span across.
+type FailureDomain struct {
+	// Id is the unique identifier of this failure domain.
```

Comment:
This comment calls ID a unique field, but there doesn't seem to be any logic or validation that the field is indeed unique.

Reply:
I'm gonna change the comment to make that the responsibility of the infrastructure provider, wdyt?

Reply:
How would that affect consumption? If we rely on the provider, then we'd still need to guard against duplication somehow on the consumption side. Either that or end up with undefined behavior.

Reply:
This is similar to what the conditions array does today. On the consumption side, I'd expect for users to assume the data is correct and leave the responsibility/bug to the infrastructure provider, if any.

Reply:
@vincepri if I'm not mistaken, conditions are expected to be consumed by taking the last condition for a given ID provided in the list. Should we define the same behavior here as well, then?

Reply:
We don't need that behavior/requirement. What I'm suggesting is from a user perspective to assume each ID and entry in the list is unique because the contract with infrastructure provider requires it to be like that.

I'll add some helper methods.